rust/src/linearview.rs

```diff
@@ -0,0 +1,515 @@
+
+        self.offset += 1;
+        Some(line)
+    }
+}
+
+
+// FIXME: we get a segmentation fault if this is there. However,
+// I don't understand enough of the API to see where this should get freed otherwise.
+/*
+impl Drop for LinearViewLinesIterator {
+    fn drop(&mut self) {
+         unsafe {
+             BNFreeLinearDisassemblyLines(self.lines, self.count);
+        }
+    }
+}
+*/
+
+pub type LinearDisassemblyLineType = BNLinearDisassemblyLineType;
+
+pub struct LinearDisassemblyLine {
+    r#type: LinearDisassemblyLineType,
+    function: Ref<Function>,
+    contents: BNDisassemblyTextLine,
+}
+
+impl LinearDisassemblyLine {
+    pub(crate) unsafe fn from_raw(raw: *mut BNLinearDisassemblyLine) -> Self {
+        let line = *raw;
+        let linetype = line.type_;
+        let function = Function::from_raw(line.function);
+        let contents = line.contents;
+        Self {
+            r#type: linetype,
+            function,
+            contents,
+        }
+    }
+
+    pub fn addr(&self) -> u64 {
+        self.contents.addr
+    }
+
+    pub fn instr_idx(&self) -> usize {
+        self.contents.instrIndex
+    }
+
+    pub fn count(&self) -> usize {
+        self.contents.count
+    }
+
+    pub fn tag_count(&self) -> usize {
+        self.contents.tagCount
+    }
+
+    pub fn tokens(&self) -> LinearDisassemblyLineTokenIterator {
+        LinearDisassemblyLineTokenIterator::new(self.contents.count, self.contents.tokens)
+    }
+
+    pub fn function(&self) -> Ref<Function> {
+        self.function.clone()
+    }
+
+    pub fn line_type(&self) -> LinearDisassemblyLineType {
+        self.r#type
+    }
+}
+
+impl std::fmt::Display for LinearDisassemblyLine {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for token in self.tokens() {
+            write!(f, "{}", token.text())?;
+        }
+
+        Ok(())
+    }
+}
+
+
+pub struct LinearDisassemblyLineTokenIterator {
+    count: usize,
+    tokens: *mut BNInstructionTextToken,
+    offset: usize,
+}
+
+impl LinearDisassemblyLineTokenIterator {
+    pub(crate) fn new(count: usize, tokens: *mut BNInstructionTextToken) -> Self {
+        Self {
+            count,
+            tokens,
+            offset: 0,
+        }
+    }
+
+    #[allow(unused)]
+    pub(crate) fn empty() -> Self {
+        Self {
+            count: 0,
+            tokens: std::ptr::null_mut(),
+            offset: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.count
+    }
+}
+
+impl Iterator for LinearDisassemblyLineTokenIterator {
+    type Item = InstructionTextToken;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.count {
+            return None;
+        }
+
+        let token = unsafe { *self.tokens.offset(self.offset as _) };
+
+        self.offset = self.offset + 1;
+        Some(InstructionTextToken(token))
+    }
+}
```

Comment:
Same note about replacing with a `CoreOwnedArrayWrapper` impl here... If you run in to problems with that, let me know.